README.md
#CourseAnalyzer
CourseAnalyzer takes the list of courses from courses.txt and downloads the raw course information from the Louisiana State University course catalog into the folder `course_cache`. 
It then proceeds to parse all of the courses and sections into objects. 
Finally, statistics (or reports) are generated into a static html website, and are merged into the gh-pages branch, available at [https://blaketillman.github.io/CourseAnalyzer](https://blaketillman.github.io/CourseAnalyzer). 

course_cache/HISTORY.txt
  5   170  HIST 1001         1  WEST CIVIL TO 1500     3.0   930-1020   M W F  0016 LOCKETT                          DIETZ M
 11   164  HIST 1001         2  WEST CIVIL TO 1500     3.0  1130-1220   M W F  0006 LOCKETT                          JOHNSON S
  9   166  HIST 1001         3  WEST CIVIL TO 1500     3.0   130-0220   M W F  0006 LOCKETT                          KOOI C
  1    39  HIST 1001         4  WEST CIVIL TO 1500     3.0   900-1020    T TH  0124CHERGET HALL                      HARDY J
  2    28  HIST 1002         1  HON: WES CIV TO 1500   3.0  1200-0120    T TH  0051 ALLEN                            HARDY J
 32   143  HIST 1003         1  HIS WES CIVILIZATION   3.0  1230-0120   M W F  0006 LOCKETT                          KARCH B
 14   128  HIST 1003         3  HIS WES CIVILIZATION   3.0  1200-0120    T TH  0009 LOCKETT                          MARTIN B
 49    75  HIST 1005         1  WORLD HIST TO 1500     3.0  1030-1120   M W F  0152 COATES                           ALAM A
 23   129  HIST 1007         1  WLD HIST SINCE 1500    3.0  1030-1120   M W F  0103 DESIGN BUILDING                  BERG S
  5   131  HIST 2055         1  US TO 1865             3.0  1230-0120   M W F  E130 HOWE RUSSELL E                   FOSTER G
 98    77  HIST 2055         2  US TO 1865             3.0   730-0850    T TH  0016 LOCKETT                          SHEEHAN-DEAN
 12   163  HIST 2055         3  US TO 1865             3.0   130-0250    T TH  0016 LOCKETT                          BURSTEIN A
  3     9  HIST 2056         1  HON: THE U. S.         3.0  1230-0120   M W F  E130 HOWE RUSSELL E                   FOSTER G
     **                    STUDENTS MUST BE AVAILABLE ON WEDNESDAYS 1:30-2:30
  3   172  HIST 2057         1  THE US: 1865 TO PRES   3.0  1130-1220   M W F  0015 LOCKETT                          SHINDO C
 10   165  HIST 2057         2  THE US: 1865 TO PRES   3.0   130-0220   M W F  0015 LOCKETT                          JACQUET C
 31   144  HIST 2057         3  THE US: 1865 TO PRES   3.0  1030-1150    T TH  0016 LOCKETT                          CARSON M
 97    78  HIST 2061         1  AFRICAN-AMERICAN HIS   3.0   130-0250    T TH  0010 LOCKETT                          ROBERTS K
      ***   HIST 2075  ***       CROSS-LISTED WITH   GERM 2075
  1    14  HIST 2075         1  GERMAN CIVILIZATION    3.0   430-0550N   T TH  0112 AUDUBON                          LEDER H
 24    16  HIST 2195         1  MODERN SOUTH AFRICA    3.0  1030-1150    T TH  0125 TUREAUD HALL                     CLARK N
 12    28  HIST 2195         2  VIOLENCE IN US WEST    3.0  1200-0120    T TH  0125 TUREAUD HALL                     GUTFREUND Z
 25    15  HIST 3001         1  HIST & SOC SCI I       1.0   300-0350    T     0138 LOCKETT                          GUTFREUND Z
 16   126  HIST 3071         1  LOUISIANA              3.0  1030-1150    T TH  0009 LOCKETT                          LONG A
  9    11  HIST 3118 SEM     1  GERM CITZS 1750 PRST   3.0   130-0250   M W    0118 PRESCOTT                         TOMLINSON G
 11     9  HIST 3119 SEM     1  ASIAN AMERICAN HIST    3.0   300-0550   M      0253 HIMES                            SHINDO C
  8    12  HIST 3119 SEM     2  BLACK PANTHERS POWER   3.0   300-0420    T TH  0240 LOCKETT                          ROBERTS K
  2    18  HIST 3119 SEM     3  AMER COLD WAR CULTUR   3.0   300-0550      TH  0108 AUDUBON                          BROKAW D
 15    25  HIST 4001         1  GREECE OF THE CITY S   3.0  1030-1150    T TH  0225 TUREAUD HALL                     ROSS S
(F)    41  HIST 4007         1  EARLY MIDDLE AGES      3.0   130-0250   M W    0236 COATES                           DIETZ M
 19    21  HIST 4016         1  EUROPE IN 19TH CEN     3.0   900-1020    T TH  0253 HIMES                            MARTIN B
 19    21  HIST 4021         1  FRANCE BEFORE 1770     3.0   130-0250    T TH  0232 LOCKETT                          TUTTLE L
 30    10  HIST 4024         1  DUTCH REPUB & EMP 15   3.0  1030-1120   M W F  0253 HIMES                            KOOI C
  4    38  HIST 4044         1  STUART ENGLAND         3.0  1230-0120   M W F  0241 LOCKETT                          STATER V
 32    28  HIST 4066         1  MILITARY HIST U S      3.0   900-1020    T TH  0109 TUREAUD HALL                     CARSON M
  1    39  HIST 4071         1  ANTEBELLUM SOUTH       3.0  1200-0120    T TH  0137 ALLEN                            SHEEHAN-DEAN
  9    31  HIST 4079         1  WOMEN AMER HISTORY     3.0  1030-1150   M W    0101 TUREAUD HALL                     JACQUET C
 15    25  HIST 4081         1  CARIBBEAN 1492-1830    3.0   130-0250    T TH  0253 HIMES                            BENSON D
 12    28  HIST 4094         1  MODERN JAPAN           3.0  1130-1220   M W F  0103 COATES                           HENDERSON J
(F)     1  HIST 4109         1  HONORS PROSEMINAR      3.0  TBA                                                      DIETZ M
 17    28  HIST 4140         1  VIETNAM WAR            3.0  1030-1150    T TH  0206 WILLIAMS                         WAGNER T
      ***   HIST 4191  ***       CROSS-LISTED WITH   REL  4191
  3    17  HIST 4191         1  RELIGNS CHINA/JAPAN    3.0   930-1020   M W F  0103 COATES                           HENDERSON J
 38     2  HIST 4195         2  AFRICA SINCE 1880      3.0  1030-1150    T TH                                        COLE G
  5    35  HIST 4197         1  CRIME COURTROOM DRAM   3.0  1200-0120    T TH  0129 TUREAUD HALL                     ISENBERG N
 30    10  HIST 4403         1  HIST & SOC SCI III     1.0   400-0450    T     0136 LOCKETT                          GUTFREUND Z
 39     1  HIST 4404 SEM     1  SEM HIST & SOC SCI     3.0   430-0720N     TH  0253 HIMES                            GUTFREUND Z
      ***   HIST 4505  ***       CROSS-LISTED WITH   REL  4505
  1    14  HIST 4505         1  RISE OF CHRISTIANITY   3.0  1030-1150    T TH  0127 TUREAUD HALL                     STORIN B

src/com/btilm305/Analyzer.java
package com.btilm305;

import com.btilm305.filter.SectionFilter;
import com.btilm305.object.ClassTime;
import com.btilm305.object.Course;
import com.btilm305.object.Section;
import com.btilm305.object.Time24Hours;
import com.btilm305.report.Report;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class Analyzer {

	public static void main(String[] args) {
		new Analyzer();
	}

	private School school = new School();
	private Section previousSection = null;

	public Analyzer() {
		try {
			long start = System.currentTimeMillis();
			File courseCacheFile = new File("course_cache");
			if (!courseCacheFile.exists()) {
				courseCacheFile.mkdirs();
				new CourseDownloader(new File("courses.txt"), courseCacheFile).run();
			}
			for (File file : courseCacheFile.listFiles()) {
				if (file.getName().endsWith(".txt")) {
					String text = Util.fileToString(file);
					for (String line : text.split("\n")) {
						parseLine(line);
					}
				}
			}
			System.out.println("Analysis took " + (System.currentTimeMillis() - start) + "ms");
			List<Section> things = new SectionFilter(school.getSections()).day(ClassTime.WEDNESDAY).time(new Time24Hours(12 + 9, 0)).filter();
			for (Section s : things) {
				System.out.println(s.getCourse().toString() + ": " + s);
			}
			Report.run(school);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private void parseLine(String line) {
		boolean additional = isAdditionSection(line);
		if (line.trim().startsWith("*")) {
			handleNote(line);
			return;
		}
		if (!additional && (looksEmpty(line))) {
			return;
		}
		line = rightPad(line, 118);
		Section section = new Section();
		section.setAvailable(line.substring(0, 3).trim());
		String enrolled = line.substring(6, 9).trim();
		section.setEnrolled(enrolled);
		section.setLab(line.substring(21, 24).equals("LAB"));
		section.setSectionNumber(line.substring(27, 30).trim());
		section.setHours(line.substring(55, 60).trim());
		String timeRange = line.substring(60, 70).trim();
		if (timeRange.trim().equals("TBA")) {
			return;
		}
		section.setTime(new ClassTime(line.substring(72, 77), timeRange));
		section.setRoom(line.substring(79, 84).trim());
		section.setBuilding(line.substring(84, 100).trim());
		section.setInstructor(line.substring(117, line.length()).trim());
		String courseAbbreviation = line.substring(11, 15).trim();
		String courseNumber = line.substring(16, 20).trim();
		String courseTitle = line.substring(32, 55).trim();
		if (additional) {
			section.setEnrolled(previousSection.getEnrolled() + "");
			section.setAvailable(previousSection.getAvailable());
			courseAbbreviation = previousSection.getCourse().getAbbreviation();
			courseNumber = previousSection.getCourse().getNumber();
			if (section.getRoom().isEmpty()) {
				section.setRoom(previousSection.getRoom());
			}
			if (section.getBuilding().equals("undefined")) {
				section.setBuilding(previousSection.getBuilding());
			}
			if (section.getInstructor().isEmpty()) {
				section.setInstructor(previousSection.getInstructor());
			}
		}

		Course course = getCourse(courseAbbreviation, courseNumber);
		if (course == null) {
			course = new Course(courseAbbreviation, courseNumber, courseTitle);
			school.getCourses().add(course);
		}
		course.addSection(section);
		school.getSections().add(section);
		if (!additional) {
			previousSection = section;
		}
	}

	private void handleNote(String line) {
		int index = line.indexOf("CLASS WILL BE HELD IN ");
		if (index != -1 && previousSection != null) {
			String location = line.substring(index + 22, line.length());
			String[] words = location.split(" ");
			String building = location;
			String room = "";
			if (words[0].matches(".*\\d+.*")) {
				room = words[0];
				building = location.substring(room.length() + 1);
			} else if (words[words.length - 1].matches(".*\\d+.*")) {
				room = words[words.length - 1];
				building = location.substring(0, location.length() - room.length() - 1);
			}
			previousSection.setBuilding(building);
			previousSection.setRoom(room);
		}
	}

	private String rightPad(String string, int length) {
		StringBuilder sb = new StringBuilder(string);
		while (sb.length() < 117) {
			sb.append(" ");
		}
		return sb.toString();
	}

	private void parseLab(String line) {
		Section section = new Section();
		section.setLab(true);
		String timeRange = line.substring(60, 70).trim();
		if (timeRange.trim().equals("TBA")) {
			return;
		}
		section.setTime(new ClassTime(line.substring(72, 77), timeRange));
		section.setRoom(line.substring(79, 84).trim());
		section.setBuilding(line.substring(84, 100).trim());
		section.setInstructor(line.substring(117, line.length()).trim());
		//fill in data from previous line
		section.setAvailable(previousSection.getAvailable());
		section.setEnrolled(previousSection.getEnrolled() + "");
		section.setSectionNumber(previousSection.getSectionNumber());
		previousSection.getCourse().addSection(section);
		school.getSections().add(section);
	}

	private boolean isAdditionSection(String line) {
		return line.length() > 70 && line.substring(0, 21).trim().isEmpty() && line.substring(24, 60).trim().isEmpty() && !line.substring(60, 70).trim().isEmpty();
	}

	private boolean looksEmpty(String line) {
		return line.length() < 117 || (line.charAt(0) == ' ' && line.charAt(1) == ' ' && line.charAt(2) == ' ') || (line.charAt(0) != ' ' && line.charAt(0) != '(' && (line.charAt(0) < '0' || line.charAt(0) > '9'));
	}

	private Course getCourse(String abbreviation, String number) {
		//using CourseFilter here would be really slow and would take up the same amount of code
		for (Course course : school.getCourses()) {
			if (course.getAbbreviation().equals(abbreviation) && course.getNumber().equals(number)) {
				return course;
			}
		}
		return null;
	}
}
